content/blog/2022/2022-08-30-handling-timezones-in-spring.md

```diff
@@ -0,0 +1,635 @@
+        "Invalid value for HourOfDay (valid values 0 - 23): 25 :=" + invalidTime);
+
+LocalTime zoneTime = LocalTime.now(ZoneId.of("Europe/London"));
+System.out.println("Zone-specific time : " + zoneTime);
+
+DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm");
+System.out.println("Formatted Date : " + zoneTime.format(formatter));
+````
+
+Output:
+
+````text
+Current Time in default format : 05:53:44.839445500
+Custom Time: 21:40:50
+Zone-specific time : 20:53:44.847446100
+Formatted Date : 20:53
+````
+
+### LocalDateTime
+
+**[java.time.LocalDateTime](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html)** is an immutable object that is a combination of both `java.time.LocalDate` and `java.time.LocalTime`.
+
+Sample Conversion examples:
+
+````java
+LocalDateTime currentDateTime = LocalDateTime.now();
+System.out.println("Current Date/Time in system default timezone : " + currentDateTime);
+
+LocalDateTime currentUsingLocals = LocalDateTime.of(LocalDate.now(), LocalTime.now());
+System.out.println("Current Date/Time with LocalDate and LocalTime in system timezone : " + currentUsingLocals);
+
+LocalDateTime customDateTime = LocalDateTime.of(2022, Month.SEPTEMBER, 1, 10, 30, 59);
+System.out.println("Custom Date/Time with custom date and custom time : " + customDateTime);
+
+DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm");
+System.out.println("Formatted Date/Time : " + LocalDateTime.now().format(formatter));
+
+LocalDateTime zoneDateTime = LocalDateTime.now(ZoneId.of("+02:00"));
+System.out.println("Zoned Date Time : " + zoneDateTime);
+
+String currentDateTimeStr = "20-02-2022 10:30:45";
+DateTimeFormatter format = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm:ss");
+System.out.println("Parsed From String to Object : " + LocalDateTime.parse(currentDateTimeStr, format));
+
+LocalDateTime invalidZoneDateTime = LocalDateTime.now(ZoneId.of("Europ/London"));
+System.out.println("Invalid Zone with Exception : java.time.zone.ZoneRulesException: " +
+        "Unknown time-zone ID: Europ/London: " + invalidZoneDateTime);
+````
+
+Output:
+````text
+Current Date/Time in system default timezone : 2022-09-12T05:54:31.941578100
+Current Date/Time with LocalDate and LocalTime in system timezone : 2022-09-12T05:54:31.945587400
+Custom Date/Time with custom date and custom time : 2022-09-01T10:30:59
+Formatted Date/Time : 12-09-2022 05:54
+Zoned Date Time : 2022-09-11T21:54:31.948621700
+Parsed From String to Object : 2022-02-20T10:30:45
+````
+
+### ZonedDateTime
+
+**[java.time.ZonedDateTime](https://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html)** is an immutable representation of date, time and timezone. It automatically handles Daylight Saving Time(DST) clock changes via the `java.time.ZoneId` which internally resolves the zone offset.
+
+Sample conversion example:
+
+````java
+ZonedDateTime currentZoneDateTime = ZonedDateTime.now();
```

Comment:
There are quite a few code examples that have not been converted to unit tests.

Reply:
Unit tests are added for all date/time classes